File: agent/query_engine/ConversationQueryEngine.py
from __future__ import annotations

from typing import AsyncIterator

from query_engine.EngineContext import EngineContext
from query_engine.EngineEvent import EngineEvent
from query_engine.EngineStrategy import EngineStrategy


class ConversationQueryEngine:
    def __init__(self, strategy: EngineStrategy) -> None:
        self._strategy = strategy

    async def query(self, context: EngineContext) -> AsyncIterator[str]:
        saw_done = False
        saw_error = False
        async for event in self._strategy.run(context):
            if event.event == "sys_done":
                saw_done = True
            if event.event == "sys_error":
                saw_error = True
            if event.trace_id is None:
                event = EngineEvent(
                    event=event.event,
                    source=event.source,
                    payload=event.payload,
                    trace_id=context.trace_id,
                    event_version=event.event_version,
                )
            yield event.to_sse()

        if not saw_done:
            finish_reason = "stream_finished_with_error" if saw_error else "stream_finished"
            yield EngineEvent.sys_done(finish_reason, trace_id=context.trace_id).to_sse()

File: agent-ts/src/query_engine/ConversationQueryEngine.ts
import type { EngineContext } from "./EngineContext.js";
import type { EngineStrategy } from "./EngineStrategy.js";
import { EngineEvent } from "./EngineEvent.js";

export class ConversationQueryEngine {
  private readonly strategy: EngineStrategy;
  private readonly progressTimeoutMs: number;

  constructor(strategy: EngineStrategy, progressTimeoutMs: number = 1000) {
    this.strategy = strategy;
    this.progressTimeoutMs = progressTimeoutMs;
  }

  async *query(context: EngineContext): AsyncIterable<string> {
    for await (const event of this.queryEvents(context)) {
      yield event.toSse();
    }
  }

  async *queryEvents(context: EngineContext): AsyncIterable<EngineEvent> {
    let sawDelta = false;
    let sawDone = false;
    let sawError = false;
    let progressSeconds = 0;
    const iterator = this.strategy.run(context)[Symbol.asyncIterator]();
    let pendingNext: Promise<IteratorResult<EngineEvent>> | null = null;

    while (true) {
      if (!pendingNext) {
        pendingNext = iterator.next();
      }
      const result = await Promise.race([
        pendingNext,
        this.createProgressTimeout()
      ]);
      if (result === "timeout") {
        if (!sawDelta) {
          progressSeconds += 1;
          yield new EngineEvent(
            "sys_progress",
            "system",
            { message: "思考模式中考量中，请稍候...", elapsed_sec: progressSeconds },
            context.traceId ?? null
          );
        }
        continue;
      }
      pendingNext = null;
      if (result.done) {
        break;
      }
      let event = result.value;
      if (event.event === "sys_done") {
        sawDone = true;
      }
      if (event.event === "sys_error") {
        sawError = true;
      }
      if (event.event === "llm_data" || event.event === "llm_delta" || event.event === "raw" || event.event === "delta") {
        sawDelta = true;
      }
      if (!event.traceId) {
        event = new EngineEvent(event.event, event.source, event.payload, context.traceId ?? null, event.eventVersion);
      }
      yield event;
    }

    if (!sawDelta && !sawDone && !sawError) {
      yield new EngineEvent("error", "system", { message: "stream finished without content" }, context.traceId ?? null);
      return;
    }
    if (!sawDone) {
      const finishReason = sawError ? "stream_finished_with_error" : "stream_finished";
      yield EngineEvent.sysDone(finishReason, context.traceId ?? null);
    }
  }

  private async createProgressTimeout(): Promise<"timeout"> {
    await new Promise((resolve) => setTimeout(resolve, this.progressTimeoutMs));
    return "timeout";
  }
}
